src/languages/scala.ts

```diff
@@ -0,0 +1,64 @@
+import {
+  createPatternMatchers,
+  argumentMatcher,
+  leadingMatcher,
+  conditionMatcher,
+  trailingMatcher,
+  cascadingMatcher,
+} from '../util/nodeMatchers';
+import { NodeMatcherAlternative, ScopeType } from '../typings/Types';
+
+const nodeMatchers: Partial<Record<ScopeType, NodeMatcherAlternative>> = {
+  // treating classes = classlike
+  class: ['class_definition', 'object_definition', 'trait_definition'],
+  className: ['class_definition[name]', 'object_definition[name]', 'trait_definition[name]'],
+
+  ifStatement: 'if_expression',
+
+  string: ['interpolated_string_expression', 'string'],
+  comment: 'comment',
+
+  // lists basic definition is just a function call to a constructor, eg List(1,2,3,4)
+  // MISSING: fancy list style: val foo = 1 :: (2 :: (3 :: Nil)) // List(1,2,3)
+  list: 'call_expression',
+  map: 'call_expression',
+
+  // list.size(), does not count foo.size (field_expression), or foo size (postfix_expression)
+  functionCall: 'call_expression',
+  namedFunction: 'function_definition',
+  // Do we want to consider partial functions as lambdas? eg
+  //   foo.map(_ + 1)
+  //   foo.map {_ + 1}
+  //   foo.map {case x => x + 1}
+  // are not under `lambda_expression`
+  anonymousFunction: 'lambda_expression',
+
+  argumentOrParameter: argumentMatcher('arguments', 'parameters', 'class_parameters', 'bindings'),
+
+  name: ['*[name]', '*[pattern]'],
+  functionName: 'function_definition[name]',
+
+  // MISSING: function/class/trait generics
+  type: leadingMatcher([
+    'generic_type.type_identifier',
+    'type_identifier',
+    'type_parameters',
+  ], [':']),
+  value: leadingMatcher(['*[value]', '*[default_value]'], ['=']),
+  condition: conditionMatcher('*[condition]'),
+
+  // Pulled from the complete list that isn't implemented above
+
+  // collectionItem: "???"
+  // collectionKey: "???",
```

Comment:
Don't you want to support `"AL" -> "Alabama"` expressions?  For these, "key" would be `"AL"`, "value" would be `"Alabama"`, and "item" would be the pair.

`collectionItem` is also supposed to include list elements, tho since those are just function calls you would basically use an argument matcher and check that the parent is a `List`.  For fancy stuff like this, you might check out the [Clojure implementation](https://github.com/pokey/cursorless-vscode/blob/main/src/languages/clojure.ts).  That one took tons of custom matchers to get working

Reply:
I think I'd have to write something custom for that: `("AL" -> "Alabama")` comes out as:
```
          arguments: (arguments [1, 19] - [1, 38]
            (infix_expression [1, 20] - [1, 37]
              left: (string [1, 20] - [1, 24])
              operator: (operator_identifier [1, 25] - [1, 27])
              right: (string [1, 28] - [1, 37]))))))))
```

So I'd need to take the left or right of an operator, when that operator === '->'

Reply:
Yeah that looks like it should be doable, but a bit custom.  I think I'd be ok with shipping without it for now but filing an issue to capture all the missing things.  I believe cursorless will generate a helpful message so that users can just see it hasn't been implemented yet

Reply:
btw I'd capture your example here into the follow-up issue for remaining scala features; it's quite helpful